barber.cpp

```diff
@@ -1,28 +1,50 @@
 #include "barber.h"
 #include "shop.h"
+#include <iostream>
+#include <chrono>
 
 Barber::Barber(Shop* shop)
-    :mAlive(true)
+    :futureObj(exitSignal.get_future())
     ,pBelongsToShop(shop)
 {
+    mBarberNotifier = shop->getBarberSemaphore();
+    mCustomersNotifier = shop->getCustomersSemaphore();
+
     mBarberThread = std::thread(operating, this);
-    mBarberThread.join();
 }
 
 void Barber::cutHair()
 {
     std::cout << "Barber : cutHair()\n";
+    std::this_thread::sleep_for(std::chrono::seconds(4));
 }
 
 void Barber::operating()
 {
-    while (mAlive)
+    while (stopRequested() == false)
     {
-        std::this_thread::sleep_for(std::chrono::seconds(1));
+        mCustomersNotifier->Wait();
+        pBelongsToShop->removeWaitingCustomer();
+        mBarberNotifier->Signal();
+        cutHair();
     }
 }
 
-void Barber::terminate()
+void Barber::joinThread()
+{
+    mBarberThread.join();
+}
+
+//Checks if thread is requested to stop
+bool Barber::stopRequested()
+{
+    // checks if value in future object is available
+    if (futureObj.wait_for(std::chrono::milliseconds(0)) == std::future_status::timeout)
```

Comment:
What is the point of future and "waiting" if you are not going to actually wait? You are still essentially checking a boolean here.

Reply:
This solution of a promise is suggested in order to terminate a thread. See the link I posted above, in my comment right after my latest push (http://thispointer.com/c11-how-to-stop-or-terminate-a-thread/). Consider that the shopo should close at some point, so the barber should stop operating. Previously you suggested that I should remove the boolean value since it was not safe and it was in general a bad pattern, and probably send a NULL object in order to stop Barber's operation.

Reply:
The future construct is thread safe, as an atomic Boolean would also be thread safe. The reason I don't like this approach either is that it is based on polling. Previously, you were polling the value of a Boolean flag. Now you are polling a future. The problem comes from the fact that the barber must react to both customers appearing and to a termination request. With the current design, he can't wait forever on one event because the other might happen. So he has to wait a little bit here (clients semaphore) and a little bit there (termination future). This checking of one channel and then checking the other is a form of polling that keeps him busy. I think, what I mean will become clear when you start running the program and you see it in action.